xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp

```diff
@@ -1800,33 +2186,16 @@ int CAMLCodec::Decode(uint8_t *pData, size_t iSize, double dts, double pts)
       m_1st_pts = am_private->am_pkt.lastpts;
   }
 
-  // if we have still frames, demux size will be small
-  // and we need to pre-buffer more.
-  double target_timesize = 1.0;
-  if (iSize < 20)
-    target_timesize = 2.0;
-
-  // keep hw buffered demux above 1 second
-  if (GetTimeSize() < target_timesize && m_speed == DVD_PLAYSPEED_NORMAL)
-    return VC_BUFFER;
-
-  // wait until we get a new frame or 25ms,
-  if (m_old_pictcnt == m_cur_pictcnt)
-    m_ready_event.WaitMSec(25);
+  if (!DequeueFrame(m_lastFrame))
+    return VC_ERROR;
 
   // we must return VC_BUFFER or VC_PICTURE,
   // default to VC_BUFFER.
   int rtn = VC_BUFFER;
-  m_player_pts = DVD_NOPTS_VALUE;
-  if (m_old_pictcnt != m_cur_pictcnt)
+  if (m_lastFrame)
   {
-    m_old_pictcnt++;
     rtn = VC_PICTURE;
-    m_player_pts = pts;
-    // we got a new pict, try and keep hw buffered demux above 2 seconds.
-    // this, combined with the above 1 second check, keeps hw buffered demux between 1 and 2 seconds.
-    // we also check to make sure we keep from filling hw buffer.
-    if (GetTimeSize() < 2.0 && GetDataSize() < m_vbufsize/3)
+    if (GetBufferLevel() < 5)
```

Comment:
How often does the hw video decoder starve and stall ? Seen by what looks like video frames are stuttering.

Really should be using buffer time here, you could have several small demux packets (black frames or frames that contain little changes), these are very small in size or several large demux packets (key frames, lots of motion, etc) and the level of buffering becomes dependent on them. To remove the variable dependency, use time. My previous tests show that if you can keep the hw buffers filled with about a second's worth of demux, it will never starve. Might be able to back it down to 0.5 secs or lower, my tests were on an M3 and currrent AML SoC's have much more ponies.

Also remember that the call into CAMLCodec::Decode happen when VideoPlayerVideo decides. So you can easily drain the hw buffers and starve the hw decoder if the amount of hw buffering is too small. Starving the hw decoder is bad, it is designed to pipeline stream, not single frame decode and goes through several state changes between starve/pause and play.